examples/hdf5.py

```diff
@@ -4,99 +4,177 @@
+
+        if isinstance(self._x, int):
+            x = np.arange(i.start, i.stop, i.step)
+        else:
+            x = self._x[i]
+
+        self._lastRange = (i1, i2, ds)
+        return x, y
+
+
+def plotDataRange(x, x1, x2, padding=0.3, sampleLimit=2500, regularize=True):
+    pad = (x2 - x1) * padding
+    x1, x2 = x1 - pad, x2 + pad
+
+    if isinstance(x, int):
+        n = x
+        i1 = max(0, min(n, int(x1)))
+        i2 = max(0, min(n, int(x2)))
+    else:
+        n = len(x)
+        i1, i2 = np.searchsorted(x, [x1, x2]).tolist()
+
+    ds = max(1, (i2 - i1) // sampleLimit)
+
+    if regularize:
+        lv = ds.bit_length()
+        ds = 2 ** lv
+        i1 = (i1 // ds - 1) * ds
+        i2 = (i2 // ds + 1) * ds
+        i1 = max(0, min(n, i1))
+        i2 = max(0, min(n, i2))
+
+    return i1, i2, ds
+
+
+def downsample(data, start, stop, ds, cache=None, **kwargs):
+    if not (0 <= start < stop <= len(data)):
+        raise ValueError
+    if ds <= 2:
+        # downsampling has no effect
+        dat = data[start:stop]
+        idx = slice(start, stop)
+    else:
+        if cache is not None and ds in cache:
+            dat = cache[ds][2 * start // ds: 2 * stop // ds + 1]
+        else:
+            dat = _downsample(data=data, start=start, stop=stop, ds=ds, **kwargs)
+        idx = slice(start, start + len(dat) * ds // 2, ds // 2)
+    return idx, dat
+
+
+def _downsample(data, start, stop, ds, chunksize=1000000):
+    # Here convert data into a down-sampled array suitable for visualizing.
+    # Must do this piecewise to limit memory usage.
+
+    samples = 1 + ((stop - start) // ds)
+    visible = np.zeros(samples * 2, dtype=data.dtype)
+    sourcePtr = start
+    targetPtr = 0
+    chunksize = (chunksize // ds) * ds
+
+    while sourcePtr < stop:
+        chunk = data[sourcePtr: min(stop, sourcePtr + chunksize)]
+        sourcePtr += len(chunk)
+
+        if len(chunk) % ds != 0:
+            # minimally extend chunk to be integral multiple of ds in a way that doesn't affect min/max
+            tail = np.full(ds - len(chunk) % ds, chunk[-1])
+            chunk = np.append(chunk, tail)
+
+        chunk = chunk.reshape(len(chunk) // ds, ds)
+
+        # compute max and min
+        chunkMax = chunk.max(axis=1)
+        chunkMin = chunk.min(axis=1)
+
+        # interleave min and max into plot data to preserve envelope shape
+        visible[targetPtr: targetPtr + chunk.shape[0] * 2: 2] = chunkMin
+        visible[1 + targetPtr: 1 + targetPtr + chunk.shape[0] * 2: 2] = chunkMax
+        targetPtr += chunk.shape[0] * 2
+
+    return visible
+
+
+def computeDownsampleCache(data, minLevelSize=2500, maxLevelSize=10000000):
+    minLevel = max(2, (len(data) // maxLevelSize).bit_length())
+    maxLevel = (len(data) // minLevelSize).bit_length()
+
+    out = {}
+    dat = None
+    for lv in range(minLevel, maxLevel + 1):
+        ds = 2 ** lv
+        if dat is None:
```

Comment:
Maybe I'm missing something here but couldn't you just call this once before the loop?

```python
levels = range(minLevel, maxLevel + 1)
lv = next(levels)
ds = 2 ** lv
out[ds] = _downsample(data=data, start=0, stop=len(data), ds=ds)

for lv in levels:
    out[2 ** lv] = _downsample(data=dat, start=0, stop=len(dat), ds=4)
       
return out
```